### PR TITLE
added providers for baidu, sina and wechat

### PR DESCRIPTION
**Pull request by zacksleo**
baidu: http://dwz.cn/
sina: http://open.weibo.com/wiki/2/short_url/shorten
wechat: https://mp.weixin.qq.com/wiki?t=resource/res_main&id=mp1443433600

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,6 +16,9 @@ This library allows you to shorten a URL, reverse is also possible.
 * [Installation](#installation)
 * [Bit.ly API](#bitly-api)
 * [Google API](#google-api)
+* [Baidu API](#baidu-api)
```
Comment (mremi) [inline README.md:19]: Can you sort by name please (and same below)?

File: tests/Mremi/UrlShortener/Tests/Provider/Baidu/BaiduProviderTest.php [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+
+        $this->mockClient($response, 'post');
+
+        $this->provider->expand($link);
+    }
+
+    /**
+     * Initializes the provider.
+     */
+    protected function setUp()
```
Comment (mremi) [inline tests/Mremi/UrlShortener/Tests/Provider/Baidu/BaiduProviderTest.php:219]: Can you move it at the top of the class please (same for `tearDown`)? By convention, these methods are declared first to increase readability.

File: tests/Mremi/UrlShortener/Tests/Provider/Sina/SinaProviderTest.php [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+
+        $this->mockClient($response, 'get');
+
+        $this->provider->expand($link);
+    }
+
+    /**
+     * Initializes the provider.
+     */
+    protected function setUp()
```
Comment (mremi) [inline tests/Mremi/UrlShortener/Tests/Provider/Sina/SinaProviderTest.php:256]: Can you move it at the top of the class please (same for `tearDown`)? By convention, these methods are declared first to increase readability.

File: tests/Mremi/UrlShortener/Tests/Provider/Wechat/WechatProviderTest.php [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+    {
+        //$this->mockClient($this->getMockResponseAsString());
+
+        $this->provider->expand($this->getBaseMockLink());
+    }
+
+    /**
+     * Initializes the provider.
+     */
+    protected function setUp()
```
Comment (mremi) [inline tests/Mremi/UrlShortener/Tests/Provider/Wechat/WechatProviderTest.php:120]: Can you move it at the top of the class please (same for `tearDown`)? By convention, these methods are declared first to increase readability.

File: tests/Mremi/UrlShortener/Tests/Provider/Wechat/WechatProviderTest.php [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+
+    /**
+     * Tests the expand method throws exception if Wechat returns a string.
+     *
+     * @expectedException        \Mremi\UrlShortener\Exception\InvalidApiResponseException
+     * @expectedExceptionMessage Wechat does not support expand url yet.
+     */
+    public function testExpandThrowsExceptionIfApiResponseIsString()
+    {
+        //$this->mockClient($this->getMockResponseAsString());
```
Comment (mremi) [inline tests/Mremi/UrlShortener/Tests/Provider/Wechat/WechatProviderTest.php:112]: Can you explain please?

**Review by mremi — COMMENTED** [review]
Thanks for this job @zacksleo , can you please fix the comments and squash your commits to have only once? Sorry for the delay.

File: tests/Mremi/UrlShortener/Tests/Provider/Wechat/WechatProviderTest.php (hunk 5 above)
Comment (zacksleo) [reply tests/Mremi/UrlShortener/Tests/Provider/Wechat/WechatProviderTest.php:112]: Wechat does not has any expand api currently, so there is no need to mock request as we does not send any request and just throw an exception. Here is the expand api actually do
```
    public function expand(LinkInterface $link, $hash = null)
    {
        throw new InvalidApiResponseException('Wechat does not support expand url yet.');
    }
```

**Comment by zacksleo** [comment]
I made a new PR to resolve the conversation in #12

**Pull request closed without merging**